crates/turborepo-lib/src/process/child.rs

```diff
@@ -0,0 +1,473 @@
+                            }
+                            debug!("waiting for child {}", pid);
+                            child.wait().await.ok()
+                        } else {
+                            None
+                        }
+                    };
+
+                    info!("starting shutdown");
+
+                    let result = tokio::time::timeout(*timeout, fut).await;
+                    match result {
+                        Ok(x) => {
+                            let exit_code = x.and_then(|es| es.code());
+                            if let Some(channel) = exit_channel {
+                                channel.send(exit_code).ok();
+                            }
+                            Ok(ChildState::Finished(exit_code))
+                        }
+                        Err(_) => {
+                            info!("graceful shutdown timed out, killing child");
+                            child.kill().await?;
+                            Ok(ChildState::Killed)
+                        }
+                    }
+                }
+
+                #[cfg(windows)]
+                {
+                    debug!("timeout not supported on windows, killing");
+                    child.kill().await?;
+                    Ok(ChildState::Killed)
+                }
+            }
+            ShutdownStyle::Kill => {
+                child.kill().await?;
+                Ok(ChildState::Killed)
+            }
+        }
+    }
+}
+
+/// A child process that can be interacted with asynchronously.
+///
+/// This is a wrapper around the `tokio::process::Child` struct, which provides
+/// a cross platform interface for spawning and managing child processes.
+#[derive(Clone, Debug)]
+pub struct Child {
+    pid: Option<u32>,
+    gid: Option<u32>,
+    state: Arc<RwLock<ChildState>>,
+    exit_channel: watch::Receiver<Option<i32>>,
+    stdin: Arc<Mutex<Option<tokio::process::ChildStdin>>>,
+    stdout: Arc<Mutex<Option<tokio::process::ChildStdout>>>,
+    stderr: Arc<Mutex<Option<tokio::process::ChildStderr>>>,
+}
+
+#[derive(Debug)]
+pub struct ChildCommandChannel(mpsc::Sender<ChildCommand>);
+
+impl ChildCommandChannel {
+    pub fn new() -> (Self, mpsc::Receiver<ChildCommand>) {
+        let (tx, rx) = mpsc::channel(1);
+        (ChildCommandChannel(tx), rx)
+    }
+
+    pub async fn kill(&self) -> Result<(), mpsc::error::SendError<ChildCommand>> {
+        self.0.send(ChildCommand::Kill).await
+    }
+
+    pub async fn stop(&self) -> Result<(), mpsc::error::SendError<ChildCommand>> {
+        self.0.send(ChildCommand::Stop).await
+    }
+}
+
+pub enum ChildCommand {
+    Stop,
+    Kill,
+}
+
+impl Child {
+    /// Start a child process, returning a handle that can be used to interact
+    /// with it. The command will be started immediately.
+    pub fn spawn(mut command: Command, shutdown_style: ShutdownStyle) -> Self {
+        let mut group = command.group().spawn().expect("failed to start child");
```

Comment:
Am I correct in understanding that `group()` sets up a separate process group, like `Setpgid` from the Go side?

Reply:
Correct. This will be more useful when we set up windows signal handling